File: src/main/java/com/qtech/bubbles/common/GraphicsProcessor.java
package com.qtech.bubbles.common;

import com.qtech.bubbles.core.utils.categories.StringUtils;

import java.awt.*;
import java.awt.font.FontRenderContext;
import java.awt.font.GlyphVector;
import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.awt.image.BufferedImageOp;
import java.awt.image.ImageObserver;
import java.awt.image.RenderedImage;
import java.awt.image.renderable.RenderableImage;
import java.text.AttributedCharacterIterator;
import java.util.List;
import java.util.Map;

@SuppressWarnings("unused")
public class GraphicsProcessor {
    private final Graphics2D gg;
    private Font fallbackFont;

    public GraphicsProcessor(Graphics2D g2d) {
        this.gg = g2d;
    }

    public GraphicsProcessor(Graphics g) {
        this((Graphics2D) g);
    }

    public void draw(Shape s) {
        gg.draw(s);
    }

    public boolean drawImage(Image img, AffineTransform xform, ImageObserver obs) {
        return gg.drawImage(img, xform, obs);
    }

    public void drawImage(BufferedImage img, BufferedImageOp op, int x, int y) {
        gg.drawImage(img, op, x, y);
    }

    public void drawRenderedImage(RenderedImage img, AffineTransform xform) {
        gg.drawRenderedImage(img, xform);
    }

    public void drawRenderableImage(RenderableImage img, AffineTransform xform) {
        gg.drawRenderableImage(img, xform);
    }

    public void drawString(String str, int x, int y) {
        gg.drawString(StringUtils.createFallbackString(str, getFont()).getIterator(), x, y);
    }

    public void drawString(String str, float x, float y) {
        gg.drawString(StringUtils.createFallbackString(str, getFont()).getIterator(), x, y);
    }

    public void drawString(AttributedCharacterIterator iterator, int x, int y) {
        gg.drawString(iterator, x, y);
    }

    public boolean drawImage(Image img, int x, int y, ImageObserver observer) {
        return gg.drawImage(img, x, y, observer);
    }

    public boolean drawImage(Image img, int x, int y, int width, int height, ImageObserver observer) {
        return gg.drawImage(img, x, y, width, height, observer);
    }

    public boolean drawImage(Image img, int x, int y, Color backgroundColor, ImageObserver observer) {
        return gg.drawImage(img, x, y, backgroundColor, observer);
    }

    public boolean drawImage(Image img, int x, int y, int width, int height, Color backgroundColor, ImageObserver observer) {
        return gg.drawImage(img, x, y, width, height, backgroundColor, observer);
    }

    public boolean drawImage(Image img, int dx1, int dy1, int dx2, int dy2, int sx1, int sy1, int sx2, int sy2, ImageObserver observer) {
        return gg.drawImage(img, dx1, dy1, dx2, dy2, sx1, sy1, sx2, sy2, observer);
    }

    public boolean drawImage(Image img, int dx1, int dy1, int dx2, int dy2, int sx1, int sy1, int sx2, int sy2, Color backgroundColor, ImageObserver observer) {
        return gg.drawImage(img, dx1, dy1, dx2, dy2, sx1, sy1, sx2, sy2, backgroundColor, observer);
    }

    public void dispose() {
        gg.dispose();
    }

    public void drawString(AttributedCharacterIterator iterator, float x, float y) {
        gg.drawString(iterator, x, y);
    }

    public void drawGlyphVector(GlyphVector g, float x, float y) {
        gg.drawGlyphVector(g, x, y);
    }

    public void fill(Shape s) {
        gg.fill(s);
    }

    public boolean hit(Rectangle rect, Shape s, boolean onStroke) {
        return gg.hit(rect, s, onStroke);
    }

    public GraphicsConfiguration getDeviceConfiguration() {
        return gg.getDeviceConfiguration();
    }

    public void setComposite(Composite comp) {
        gg.setComposite(comp);
    }

    public void setPaint(Paint paint) {
        gg.setPaint(paint);
    }

    public void setStroke(Stroke s) {
        gg.setStroke(s);
    }

    public void setRenderingHint(RenderingHints.Key hintKey, Object hintValue) {
        gg.setRenderingHint(hintKey, hintValue);
    }

    public Object getRenderingHint(RenderingHints.Key hintKey) {
        return gg.getRenderingHint(hintKey);
    }

    public void setRenderingHints(Map<?, ?> hints) {
        gg.setRenderingHints(hints);
    }

    public void addRenderingHints(Map<?, ?> hints) {
        gg.addRenderingHints(hints);
    }

    public RenderingHints getRenderingHints() {
        return gg.getRenderingHints();
    }

    public GraphicsProcessor create() {
        return new GraphicsProcessor((Graphics2D) gg.create());
    }

    public void translate(int x, int y) {
        gg.translate(x, y);
    }

    public Color getColor() {
        return gg.getColor();
    }

    public void setColor(Color c) {
        gg.setColor(c);
    }

    public void setPaintMode() {
        gg.setPaintMode();
    }

    public void setXORMode(Color c1) {
        gg.setXORMode(c1);
    }

    public Font getFont() {
        return gg.getFont();
    }

    public void setFont(Font font) {
        gg.setFont(font);
    }

    public FontMetrics getFontMetrics(Font f) {
        return gg.getFontMetrics(f);
    }

    public Rectangle getClipBounds() {
        return gg.getClipBounds();
    }

    public void clipRect(int x, int y, int width, int height) {
        gg.clipRect(x, y, width, height);
    }

    public void setClip(int x, int y, int width, int height) {
        gg.setClip(x, y, width, height);
    }

    public Shape getClip() {
        return gg.getClip();
    }

    public void setClip(Shape clip) {
        gg.setClip(clip);
    }

    public void copyArea(int x, int y, int width, int height, int dx, int dy) {
        gg.copyArea(x, y, width, height, dx, dy);
    }

    public void drawLine(int x1, int y1, int x2, int y2) {
        gg.drawLine(x1, y1, x2, y2);
    }

    public void fillRect(int x, int y, int width, int height) {
        gg.fillRect(x, y, width, height);
    }

    public void clearRect(int x, int y, int width, int height) {
        gg.clearRect(x, y, width, height);
    }

    public void drawRoundRect(int x, int y, int width, int height, int arcWidth, int arcHeight) {
        gg.drawRoundRect(x, y, width, height, arcWidth, arcHeight);
    }

    public void fillRoundRect(int x, int y, int width, int height, int arcWidth, int arcHeight) {
        gg.fillRoundRect(x, y, width, height, arcWidth, arcHeight);
    }

    public void drawOval(int x, int y, int width, int height) {
        gg.drawOval(x, y, width, height);
    }

    public void fillOval(int x, int y, int width, int height) {
        gg.fillOval(x, y, width, height);
    }

    public void drawArc(int x, int y, int width, int height, int startAngle, int arcAngle) {
        gg.drawArc(x, y, width, height, startAngle, arcAngle);
    }

    public void fillArc(int x, int y, int width, int height, int startAngle, int arcAngle) {
        gg.fillArc(x, y, width, height, startAngle, arcAngle);
    }

    public void drawPolyline(int[] xPoints, int[] yPoints, int nPoints) {
        gg.drawPolyline(xPoints, yPoints, nPoints);
    }

    public void drawPolygon(int[] xPoints, int[] yPoints, int nPoints) {
        gg.drawPolygon(xPoints, yPoints, nPoints);
    }

    public void fillPolygon(int[] xPoints, int[] yPoints, int nPoints) {
        gg.fillPolygon(xPoints, yPoints, nPoints);
    }

    public void translate(double tx, double ty) {
        gg.translate(tx, ty);
    }

    public void rotate(double theta) {
        gg.rotate(theta);
    }

    public void rotate(double theta, double x, double y) {
        gg.rotate(theta, x, y);
    }

    public void scale(double sx, double sy) {
        gg.scale(sx, sy);
    }

    public void shear(double shx, double shy) {
        gg.shear(shx, shy);
    }

    public void transform(AffineTransform Tx) {
        gg.transform(Tx);
    }

    public void setTransform(AffineTransform Tx) {
        gg.setTransform(Tx);
    }

    public AffineTransform getTransform() {
        return gg.getTransform();
    }

    public Paint getPaint() {
        return gg.getPaint();
    }

    public Composite getComposite() {
        return gg.getComposite();
    }

    public void setBackground(Color color) {
        gg.setBackground(color);
    }

    public Color getBackground() {
        return gg.getBackground();
    }

    public Stroke getStroke() {
        return gg.getStroke();
    }

    public void clip(Shape s) {
        gg.clip(s);
    }

    public FontRenderContext getFontRenderContext() {
        return gg.getFontRenderContext();
    }

    public void setFallbackFont(Font font) {
        fallbackFont = font;
    }

    public Font getFallbackFont() {
        return fallbackFont;
    }

    public void drawMultiLineString(String str, int x, int y) {
        y -= gg.getFontMetrics().getHeight();

        for (String line : str.split("\n"))
            drawString(line, x, y += gg.getFontMetrics().getHeight());
    }

    public void drawWrappedString(String str, int x, int y, int maxWidth) {
        List<String> lines = StringUtils.wrap(str, getFontMetrics(getFont()), maxWidth);
        String joined = org.apache.commons.lang3.StringUtils.join(lines.toArray(new String[]{}), '\n');
//        System.out.println(joined.charAt(0));
        drawMultiLineString(joined, x, y);
    }

    public void drawTabString(String str, int x, int y) {
        for (String line : str.split("\t"))
            drawString(line, x += gg.getFontMetrics().getHeight(), y);
    }

    public GraphicsProcessor create(int x, int y, int width, int height) {
        return new GraphicsProcessor((Graphics2D) gg.create(x, y, width, height));
    }
    
    public Graphics2D getWrapped() {
        return gg;
    }

    public void fillPolygon(Polygon polygon) {
        gg.fillPolygon(polygon);
    }

    public void drawPolygon(Polygon polygon) {
        gg.drawPolygon(polygon);
    }

    public void drawRect(int x, int y, int width, int height) {
        gg.drawRect(x, y, width, height);
    }
}

File: src/main/kotlin/com/qtech/bubbles/graphics/GraphicsProcessor.kt
@file:Suppress("unused")

package com.qtech.bubbles.graphics

import com.qtech.bubbles.BubbleBlaster
import com.qtech.bubbles.core.utils.categories.GraphicsUtils.drawCenteredString
import com.qtech.bubbles.core.utils.categories.GraphicsUtils.drawLeftAnchoredString
import com.qtech.bubbles.core.utils.categories.GraphicsUtils.drawRightAnchoredString
import com.qtech.bubbles.core.utils.categories.StringUtils.createFallbackString
import com.qtech.bubbles.core.utils.categories.StringUtils.splitIntoLines
import com.qtech.bubbles.core.utils.categories.StringUtils.wrap
import com.qtech.bubbles.graphics.shapes.Circle
import com.qtech.bubbles.graphics.shapes.CircleDouble
import com.qtech.bubbles.graphics.shapes.CircleFloat
import org.apache.commons.lang3.StringUtils
import java.awt.*
import java.awt.font.FontRenderContext
import java.awt.font.GlyphVector
import java.awt.geom.AffineTransform
import java.awt.geom.Ellipse2D
import java.awt.image.BufferedImage
import java.awt.image.BufferedImageOp
import java.awt.image.ImageObserver
import java.awt.image.RenderedImage
import java.awt.image.renderable.RenderableImage
import java.text.AttributedCharacterIterator

class GraphicsProcessor {
    private val gg: Graphics2D
    private var fallbackFont: Font? = null

    constructor(g2d: Graphics2D) {
        gg = g2d
    }

    constructor(g: Graphics) {
        gg = g as Graphics2D
    }

    constructor(gg: GraphicsProcessor) {
        this.gg = gg.wrapped()
        fallbackFont = gg.fallbackFont
    }

    fun img(img: Image?, xform: AffineTransform?, obs: ImageObserver?): Boolean {
        return gg.drawImage(img, xform, obs)
    }

    fun img(img: BufferedImage?, op: BufferedImageOp?, x: Int, y: Int): GraphicsProcessor {
        gg.drawImage(img, op, x, y)
        return this
    }

    fun renderedImg(img: RenderedImage?, xform: AffineTransform?): GraphicsProcessor {
        gg.drawRenderedImage(img, xform)
        return this
    }

    fun renderableImg(img: RenderableImage?, xform: AffineTransform?): GraphicsProcessor {
        gg.drawRenderableImage(img, xform)
        return this
    }

    fun textI(str: String?, x: Int, y: Int): GraphicsProcessor {
        gg.drawString(createFallbackString(str!!, font()).iterator, x, y)
        return this
    }

    fun textF(str: String?, x: Float, y: Float): GraphicsProcessor {
        gg.drawString(createFallbackString(str!!, font()).iterator, x, y)
        return this
    }

    fun textI(iterator: AttributedCharacterIterator?, x: Int, y: Int): GraphicsProcessor {
        gg.drawString(iterator, x, y)
        return this
    }

    fun textF(iterator: AttributedCharacterIterator?, x: Float, y: Float): GraphicsProcessor {
        gg.drawString(iterator, x, y)
        return this
    }

    fun img(img: Image?, x: Int, y: Int): GraphicsProcessor {
        gg.drawImage(img, x, y, BubbleBlaster.instance)
        return this
    }

    fun img(img: Image?, x: Int, y: Int, width: Int, height: Int): GraphicsProcessor {
        gg.drawImage(img, x, y, width, height, BubbleBlaster.instance)
        return this
    }

    fun img(img: Image?, x: Int, y: Int, backgroundColor: Color?): GraphicsProcessor {
        gg.drawImage(img, x, y, backgroundColor, BubbleBlaster.instance)
        return this
    }

    fun img(img: Image?, x: Int, y: Int, width: Int, height: Int, backgroundColor: Color?): GraphicsProcessor {
        gg.drawImage(img, x, y, width, height, backgroundColor, BubbleBlaster.instance)
        return this
    }

    fun img(img: Image?, dx1: Int, dy1: Int, dx2: Int, dy2: Int, sx1: Int, sy1: Int, sx2: Int, sy2: Int): GraphicsProcessor {
        gg.drawImage(img, dx1, dy1, dx2, dy2, sx1, sy1, sx2, sy2, BubbleBlaster.instance)
        return this
    }

    fun img(img: Image?, dx1: Int, dy1: Int, dx2: Int, dy2: Int, sx1: Int, sy1: Int, sx2: Int, sy2: Int, backgroundColor: Color?): GraphicsProcessor {
        gg.drawImage(img, dx1, dy1, dx2, dy2, sx1, sy1, sx2, sy2, backgroundColor, BubbleBlaster.instance)
        return this
    }

    fun dispose(): GraphicsProcessor {
        gg.dispose()
        return this
    }

    fun glyph(g: GlyphVector?, x: Float, y: Float): GraphicsProcessor {
        gg.drawGlyphVector(g, x, y)
        return this
    }

    fun shape(s: Shape?): GraphicsProcessor {
        gg.draw(s)
        return this
    }

    fun shapeF(s: Shape?): GraphicsProcessor {
        gg.fill(s)
        return this
    }

    fun hit(rect: Rectangle?, s: Shape?, onStroke: Boolean): Boolean {
        return gg.hit(rect, s, onStroke)
    }

    val deviceConfiguration: GraphicsConfiguration
        get() = gg.deviceConfiguration

    fun composite(): Composite {
        return gg.composite
    }

    fun composite(comp: Composite?): GraphicsProcessor {
        gg.composite = comp
        return this
    }

    fun paint(): Paint {
        return gg.paint
    }

    fun paint(paint: Paint?): GraphicsProcessor {
        gg.paint = paint
        return this
    }

    fun stroke(): Stroke {
        return gg.stroke
    }

    fun stroke(s: Stroke?): GraphicsProcessor {
        gg.stroke = s
        return this
    }

    fun hint(hintKey: RenderingHints.Key?, hintValue: Any?): GraphicsProcessor {
        gg.setRenderingHint(hintKey, hintValue)
        return this
    }

    fun hint(hintKey: RenderingHints.Key?): Any {
        return gg.getRenderingHint(hintKey)
    }

    fun hints(hints: Map<*, *>?): GraphicsProcessor {
        gg.setRenderingHints(hints)
        return this
    }

    fun addHint(hints: Map<*, *>?): GraphicsProcessor {
        gg.addRenderingHints(hints)
        return this
    }

    fun hints(): RenderingHints {
        return gg.renderingHints
    }

    fun create(): GraphicsProcessor {
        return GraphicsProcessor(gg.create())
    }

    fun translate(x: Int, y: Int): GraphicsProcessor {
        gg.translate(x, y)
        return this
    }

    fun color(): Color {
        return gg.color
    }

    fun color(c: Color?): GraphicsProcessor {
        gg.color = c
        return this
    }

    fun paintMode(): GraphicsProcessor {
        gg.setPaintMode()
        return this
    }

    fun xorMode(c1: Color?): GraphicsProcessor {
        gg.setXORMode(c1)
        return this
    }

    fun background(): Color {
        return gg.background
    }

    fun background(color: Color?): GraphicsProcessor {
        gg.background = color
        return this
    }

    fun font(): Font {
        return gg.font
    }

    fun font(font: Font?): GraphicsProcessor {
        gg.font = font
        return this
    }

    fun fontMetrics(f: Font?): FontMetrics {
        return gg.getFontMetrics(f)
    }

    fun fontMetrics(): FontMetrics {
        return gg.fontMetrics
    }

    fun clipBounds(): Rectangle {
        return gg.clipBounds
    }

    fun clipRect(x: Int, y: Int, width: Int, height: Int): GraphicsProcessor {
        gg.clipRect(x, y, width, height)
        return this
    }

    fun clip1(x: Int, y: Int, width: Int, height: Int): GraphicsProcessor {
        gg.setClip(x, y, width, height)
        return this
    }

    fun clip(): Shape {
        return gg.clip
    }

    fun clip1(clip: Shape?): GraphicsProcessor {
        gg.clip = clip
        return this
    }

    fun copyArea(x: Int, y: Int, width: Int, height: Int, dx: Int, dy: Int): GraphicsProcessor {
        gg.copyArea(x, y, width, height, dx, dy)
        return this
    }

    fun line(x1: Int, y1: Int, x2: Int, y2: Int): GraphicsProcessor {
        gg.drawLine(x1, y1, x2, y2)
        return this
    }

    fun rect(x1: Int, y1: Int, x2: Int, y2: Int): GraphicsProcessor {
        gg.drawRect(x1, y1, x2, y2)
        return this
    }

    fun rectF(x: Int, y: Int, width: Int, height: Int): GraphicsProcessor {
        gg.fillRect(x, y, width, height)
        return this
    }

    fun clear(x: Int, y: Int, width: Int, height: Int): GraphicsProcessor {
        gg.clearRect(x, y, width, height)
        return this
    }

    fun roundRect(x: Int, y: Int, width: Int, height: Int, arcWidth: Int, arcHeight: Int): GraphicsProcessor {
        gg.drawRoundRect(x, y, width, height, arcWidth, arcHeight)
        return this
    }

    fun roundRectF(x: Int, y: Int, width: Int, height: Int, arcWidth: Int, arcHeight: Int): GraphicsProcessor {
        gg.fillRoundRect(x, y, width, height, arcWidth, arcHeight)
        return this
    }

    fun oval(x: Int, y: Int, width: Int, height: Int): GraphicsProcessor {
        gg.drawOval(x, y, width, height)
        return this
    }

    fun ovalF(x: Int, y: Int, width: Int, height: Int): GraphicsProcessor {
        gg.fillOval(x, y, width, height)
        return this
    }

    fun oval(x: Float, y: Float, width: Float, height: Float): GraphicsProcessor {
        gg.draw(Ellipse2D.Float(x, y, width, height))
        return this
    }

    fun ovalF(x: Float, y: Float, width: Float, height: Float): GraphicsProcessor {
        gg.fill(Ellipse2D.Float(x, y, width, height))
        return this
    }

    fun oval(x: Double, y: Double, width: Double, height: Double): GraphicsProcessor {
        gg.draw(Ellipse2D.Double(x, y, width, height))
        return this
    }

    fun ovalF(x: Double, y: Double, width: Double, height: Double): GraphicsProcessor {
        gg.fill(Ellipse2D.Double(x, y, width, height))
        return this
    }

    fun circ(x: Int, y: Int, radius: Int): GraphicsProcessor {
        gg.draw(Ellipse2D.Float(x.toFloat(), y.toFloat(), radius.toFloat(), radius.toFloat()))
        return this
    }

    fun circF(x: Int, y: Int, radius: Int): GraphicsProcessor {
        gg.fill(Ellipse2D.Float(x.toFloat(), y.toFloat(), radius.toFloat(), radius.toFloat()))
        return this
    }

    fun circ(x: Float, y: Float, radius: Float): GraphicsProcessor {
        gg.draw(Ellipse2D.Float(x, y, radius, radius))
        return this
    }

    fun circF(x: Float, y: Float, radius: Float): GraphicsProcessor {
        gg.fill(Ellipse2D.Float(x, y, radius, radius))
        return this
    }

    fun circ(x: Double, y: Double, radius: Double): GraphicsProcessor {
        gg.draw(Ellipse2D.Double(x, y, radius, radius))
        return this
    }

    fun circF(x: Double, y: Double, radius: Double): GraphicsProcessor {
        gg.fill(Ellipse2D.Double(x, y, radius, radius))
        return this
    }

    fun circ(circle: Circle?): GraphicsProcessor {
        gg.draw(circle)
        return this
    }

    fun circF(circle: Circle?): GraphicsProcessor {
        gg.fill(circle)
        return this
    }

    fun circ(circle: CircleFloat?): GraphicsProcessor {
        gg.draw(circle)
        return this
    }

    fun circF(circle: CircleFloat?): GraphicsProcessor {
        gg.fill(circle)
        return this
    }

    fun circ(circle: CircleDouble?): GraphicsProcessor {
        gg.draw(circle)
        return this
    }

    fun circF(circle: CircleDouble?): GraphicsProcessor {
        gg.fill(circle)
        return this
    }

    fun arc(x: Int, y: Int, width: Int, height: Int, startAngle: Int, arcAngle: Int): GraphicsProcessor {
        gg.drawArc(x, y, width, height, startAngle, arcAngle)
        return this
    }

    fun arcF(x: Int, y: Int, width: Int, height: Int, startAngle: Int, arcAngle: Int): GraphicsProcessor {
        gg.fillArc(x, y, width, height, startAngle, arcAngle)
        return this
    }

    fun polyLine(xPoints: IntArray?, yPoints: IntArray?, nPoints: Int): GraphicsProcessor {
        gg.drawPolyline(xPoints, yPoints, nPoints)
        return this
    }

    fun poly(polygon: Polygon?): GraphicsProcessor {
        gg.drawPolygon(polygon)
        return this
    }

    fun poly(xPoints: IntArray?, yPoints: IntArray?, nPoints: Int): GraphicsProcessor {
        gg.drawPolygon(xPoints, yPoints, nPoints)
        return this
    }

    fun polyF(polygon: Polygon?): GraphicsProcessor {
        gg.fillPolygon(polygon)
        return this
    }

    fun polyF(xPoints: IntArray?, yPoints: IntArray?, nPoints: Int): GraphicsProcessor {
        gg.fillPolygon(xPoints, yPoints, nPoints)
        return this
    }

    fun translate(tx: Double, ty: Double): GraphicsProcessor {
        gg.translate(tx, ty)
        return this
    }

    fun rotate(theta: Double): GraphicsProcessor {
        gg.rotate(theta)
        return this
    }

    fun rotate(theta: Double, x: Double, y: Double): GraphicsProcessor {
        gg.rotate(theta, x, y)
        return this
    }

    fun scale(sx: Double, sy: Double): GraphicsProcessor {
        gg.scale(sx, sy)
        return this
    }

    fun shear(shx: Double, shy: Double): GraphicsProcessor {
        gg.shear(shx, shy)
        return this
    }

    fun transform(Tx: AffineTransform?): GraphicsProcessor {
        gg.transform(Tx)
        return this
    }

    fun transform1(Tx: AffineTransform?): GraphicsProcessor {
        gg.transform = Tx
        return this
    }

    fun transform(): AffineTransform {
        return gg.transform
    }

    fun clip(s: Shape?): GraphicsProcessor {
        gg.clip(s)
        return this
    }

    val fontRenderContext: FontRenderContext
        get() = gg.fontRenderContext

    fun fallbackFont(font: Font?): GraphicsProcessor {
        fallbackFont = font
        return this
    }

    fun fallbackFont(): Font? {
        return fallbackFont
    }

    @Deprecated("")
    fun multilineText(str: String, x: Int, y: Int): GraphicsProcessor {
        var y1 = y
        y1 -= gg.fontMetrics.height
        for (line in str.split("\n").toTypedArray()) textI(line, x, gg.fontMetrics.height.let { y1 += it; y1 })
        return this
    }

    @Suppress("DEPRECATION")
    @Deprecated("")
    fun wrappedText(str: String?, x: Int, y: Int, maxWidth: Int): GraphicsProcessor {
        val lines = wrap(str!!, fontMetrics(font()), maxWidth)
        val joined: String = StringUtils.join(lines.toTypedArray(), '\n')
        multilineText(joined, x, y)
        return this
    }

    @Deprecated("")
    fun tabText(str: String, x: Int, y: Int): GraphicsProcessor {
        var x1 = x
        for (line in str.split("\t").toTypedArray()) textI(line, gg.fontMetrics.stringWidth("    ").let { x1 += it; x1 }, y)
        return this
    }

    @Deprecated("")
    fun leftText(text: String?, pnt: Point?, height: Int): GraphicsProcessor {
        drawLeftAnchoredString(this, text, pnt!!, height.toDouble(), font())
        return this
    }

    @Deprecated("")
    fun centerText(text: String?, rect: Rectangle?): GraphicsProcessor {
        drawCenteredString(this, text!!, rect!!, font())
        return this
    }

    @Deprecated("")
    fun rightText(text: String?, pnt: Point?, height: Int): GraphicsProcessor {
        drawRightAnchoredString(this, text!!, pnt!!, height.toDouble(), font())
        return this
    }

    fun centerText(text: String, x: Int, y: Int, multiline: Boolean, maxWidth: Int, wrap: Boolean, tab: Boolean, anchor: Anchor?): GraphicsProcessor {
        var x1 = x
        var y1 = y
        val metrics = fontMetrics(font())
        return when {
            wrap -> {
                val lines = wrap(text, fontMetrics(font()), maxWidth)
                val joined: String = StringUtils.join(lines.toTypedArray(), '\n')
                if (multiline) {
                    y1 -= gg.fontMetrics.height
                    for (line in splitIntoLines(joined)) centerText(line, x1, gg.fontMetrics.height.let { y1 += it; y1 }, false, maxWidth, false, tab, anchor)
                } else {
                    y1 -= gg.fontMetrics.height
                    for (line in lines) centerText(line, x1, gg.fontMetrics.height.let { y1 += it; y1 }, false, maxWidth, false, tab, anchor)
                }
                this
            }
            multiline -> {
                y1 -= gg.fontMetrics.height
                for (line in splitIntoLines(text)) centerText(line, x1, gg.fontMetrics.height.let { y1 += it; y1 }, false, maxWidth, false, tab, anchor)
                this
            }
            tab -> {
                for (line in text.split("\t").toTypedArray()) centerText(line, gg.fontMetrics.charWidth(' ') * (gg.fontMetrics.maxAdvance * 4).let { x1 += it; x1 }, y1, false, maxWidth, wrap = false, tab = false, anchor = anchor)
                this
            }
            else -> {
                val x2 = x1 + metrics.stringWidth(text) / 2
                val y2 = y1 + metrics.height / 2 + metrics.ascent
                textI(text, x2, y2)
                this
            }
        }
    }

    fun create(x: Int, y: Int, width: Int, height: Int): GraphicsProcessor {
        return GraphicsProcessor(gg.create(x, y, width, height))
    }

    fun create(rect: Rectangle): GraphicsProcessor {
        return GraphicsProcessor(gg.create(rect.x, rect.y, rect.width, rect.height))
    }

    fun wrapped(): Graphics2D {
        return gg
    }
}